Restrictive `AssetPath` `From` impls cause hard-to-diagnose compile errors

## Bevy version

0.12 or git

## What you did

Use `AssetServer` methods that take `impl Into<AssetPath<'a>>` as parameters.

## What went wrong

```
error[E0597]: `fileset` does not live long enough
   --> crates/artifice_editor/src/files.rs:120:30
    |
96  |         .show(egui_context, |ui| {
    |                             ---- value captured here
...
120 |                 for path in &fileset.files {
    |                              ^^^^^^^ borrowed value does not live long enough
...
133 |                         let handle = server.load_untyped(relpath);
    |                                      ---------------------------- argument requires that `fileset` is borrowed for `'static`
...
145 | }
    |  - `fileset` dropped here while still borrowed

error[E0521]: borrowed data escapes outside of function
   --> crates/artifice_editor/src/files.rs:133:38
    |
55  |     mut fileset: Local<FileSet>,
    |     -----------
    |     |
    |     `fileset` is a reference that is only valid in the function body
    |     has type `bevy::prelude::Local<'1, FileSet>`
...
133 |                         let handle = server.load_untyped(relpath);
    |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                                      |
    |                                      `fileset` escapes the function body here
    |                                      argument requires that `'1` must outlive `'static`
```

This diagnostic is difficult to decipher: the `impl Into<AssetPath<'a>>` argument makes it *look* like the method wouldn't have any lifetime constraints, and the diagnostics don't refer to the actual reason those constraints exist (so arguably this is a rustc bug, but bevy might want to fix it too, since what it does is somewhat odd here).

The real cause of these errors is that `AssetPath` has several `From` impls that are restricted to `'static` lifetimes:

- `impl From<&'static Path> for AssetPath<'static>`
- `impl From<&'static str> for AssetPath<'static>`

Those are chosen by rustc and are what cause the argument to be forced to `&'static Path`, instead of `&'a Path`.

It is possible to fix this by making the `From` impls fully generic, and by moving the "static lifetime optimization" to dedicated methods (right now the `AssetPath` API works exactly the other way around).

--- a/crates/bevy_asset/src/io/source.rs
+++ b/crates/bevy_asset/src/io/source.rs
@@ -70,9 +70,26 @@ impl<'a> AssetSourceId<'a> {
     }
 }
 
-impl From<&'static str> for AssetSourceId<'static> {
-    fn from(value: &'static str) -> Self {
-        AssetSourceId::Name(value.into())
+impl AssetSourceId<'static> {
+    /// Indicates this [`AssetSourceId`] should have a static lifetime.
+    #[inline]
+    pub fn as_static(self) -> Self {
+        match self {
+            Self::Default => Self::Default,
+            Self::Name(value) => Self::Name(value.as_static()),
+        }
+    }
+
+    /// Constructs an [`AssetSourceId`] with a static lifetime.
+    #[inline]
+    pub fn from_static(value: impl Into<Self>) -> Self {
+        value.into().as_static()
+    }
+}
+
+impl<'a> From<&'a str> for AssetSourceId<'a> {
+    fn from(value: &'a str) -> Self {
+        AssetSourceId::Name(CowArc::Borrowed(value))
     }
 }
 
@@ -82,10 +99,10 @@ impl<'a, 'b> From<&'a AssetSourceId<'b>> for AssetSourceId<'b> {
     }
 }
 
-impl From<Option<&'static str>> for AssetSourceId<'static> {
-    fn from(value: Option<&'static str>) -> Self {
+impl<'a> From<Option<&'a str>> for AssetSourceId<'a> {
+    fn from(value: Option<&'a str>) -> Self {
         match value {
-            Some(value) => AssetSourceId::Name(value.into()),
+            Some(value) => AssetSourceId::Name(CowArc::Borrowed(value)),
             None => AssetSourceId::Default,
         }
     }
@@ -302,7 +319,7 @@ pub struct AssetSourceBuilders {
 impl AssetSourceBuilders {
     /// Inserts a new builder with the given `id`
     pub fn insert(&mut self, id: impl Into<AssetSourceId<'static>>, source: AssetSourceBuilder) {
-        match id.into() {
+        match AssetSourceId::from_static(id) {
             AssetSourceId::Default => {
                 self.default = Some(source);
             }

--- a/crates/bevy_asset/src/lib.rs
+++ b/crates/bevy_asset/src/lib.rs
@@ -341,7 +341,7 @@ impl AssetApp for App {
         id: impl Into<AssetSourceId<'static>>,
         source: AssetSourceBuilder,
     ) -> &mut Self {
-        let id = id.into();
+        let id = AssetSourceId::from_static(id);
         if self.world().get_resource::<AssetServer>().is_some() {
             error!("{} must be registered before `AssetPlugin` (typically added as part of `DefaultPlugins`)", id);
         }

--- a/crates/bevy_asset/src/path.rs
+++ b/crates/bevy_asset/src/path.rs
@@ -475,14 +475,43 @@ impl<'a> AssetPath<'a> {
     }
 }
 
-impl From<&'static str> for AssetPath<'static> {
+impl AssetPath<'static> {
+    /// Indicates this [`AssetPath`] should have a static lifetime.
     #[inline]
-    fn from(asset_path: &'static str) -> Self {
+    pub fn as_static(self) -> Self {
+        let Self {
+            source,
+            path,
+            label,
+        } = self;
+
+        let source = source.as_static();
+        let path = path.as_static();
+        let label = label.map(CowArc::as_static);
+
+        Self {
+            source,
+            path,
+            label,
+        }
+    }
+
+    /// Constructs an [`AssetPath`] with a static lifetime.
+    #[inline]
+    pub fn from_static(value: impl Into<Self>) -> Self {
+        value.into().as_static()
+    }
+}
+
+impl<'a> From<&'a str> for AssetPath<'a> {
+    #[inline]
+    fn from(asset_path: &'a str) -> Self {
         let (source, path, label) = Self::parse_internal(asset_path).unwrap();
+
         AssetPath {
             source: source.into(),
-            path: CowArc::Static(path),
-            label: label.map(CowArc::Static),
+            path: CowArc::Borrowed(path),
+            label: label.map(CowArc::Borrowed),
         }
     }
 }
@@ -501,12 +530,12 @@ impl From<String> for AssetPath<'static> {
     }
 }
 
-impl From<&'static Path> for AssetPath<'static> {
+impl<'a> From<&'a Path> for AssetPath<'a> {
     #[inline]
-    fn from(path: &'static Path) -> Self {
+    fn from(path: &'a Path) -> Self {
         Self {
             source: AssetSourceId::Default,
-            path: CowArc::Static(path),
+            path: CowArc::Borrowed(path),
             label: None,
         }
     }

--- a/crates/bevy_utils/src/cow_arc.rs
+++ b/crates/bevy_utils/src/cow_arc.rs
@@ -26,6 +26,18 @@ pub enum CowArc<'a, T: ?Sized + 'static> {
     Owned(Arc<T>),
 }
 
+impl<T: ?Sized> CowArc<'static, T> {
+    /// Indicates this [`CowArc`] should have a static lifetime.
+    /// This ensures if this was created with a value `Borrowed(&'static T)`, it is replaced with `Static(&'static T)`.
+    #[inline]
+    pub fn as_static(self) -> Self {
+        match self {
+            Self::Borrowed(value) | Self::Static(value) => Self::Static(value),
+            Self::Owned(value) => Self::Owned(value),
+        }
+    }
+}
+
 impl<'a, T: ?Sized> Deref for CowArc<'a, T> {
     type Target = T;
 

